rir/src/compiler/pir/instruction.h

```diff
@@ -125,14 +128,49 @@ class Instruction : public Value {
         unsigned idx() const { return second; }
     };
 
-    Instruction(Tag tag, PirType t, unsigned srcIdx)
-        : Value(t, tag), srcIdx(srcIdx) {}
+    Instruction(Tag tag, PirType t, Effects effects, unsigned srcIdx)
+        : Value(t, tag), effects(effects), srcIdx(srcIdx) {}
+
+    Effects effects;
+
+    bool hasEffect() const { return !effects.empty(); }
+
+    Effects getObservableEffects() const {
+        auto e = effects;
+        // Yes visibility is a global effect. We try to preserve it. But geting
+        // it wrong is not a strong correctness issue.
+        e.reset(Effect::Visibility);
+        e.reset(Effect::Order);
+        return e;
+    }
+
+    bool isDeoptBarrier() const { return !getObservableEffects().empty(); }
+    bool mightChangeVisibility() const {
+        return effects.includes(Effect::Visibility);
+    }
+    bool mayUseReflection() const {
+        return effects.includes(Effect::Reflection);
+    }
+    bool mayForcePromises() const { return effects.includes(Effect::Force); }
+    bool leaksArg(Value* val) const {
+        // TODO: for escape analysis we use leaksEnv || hasEffect as a very
+        // crude approximation whether this instruction leaks arguments. We
+        // should do better.
+        return leaksEnv() || effects.includes(Effect::LeakArg);
+    }
+
+    void maskEffectsOnNonObjects() {
```

Comment:
This is only for binops/unops. The name mask also confused me. I would put sth like resetBinopEffects and add the `(maybeObj) Effects::any()`

Reply:
Also why the close the if with `};`

Reply:
mask as in masking a bitset

Reply:
(which the function was not doing :) so I fixed that part)